utopia-remix/app/util/github-branch-contents.server.ts

```diff
@@ -0,0 +1,345 @@
+      zipFilePath: zipFilePath,
+      existingAssets: params.existingAssets,
+    })
+
+    // 6. if there are any assets to upload, upload them
+    if (params.uploadAssets) {
+      await uploadAssets({
+        assets: assetsToUpload,
+        projectId: params.projectId,
+      })
+    }
+
+    return toApiSuccess({
+      branch: {
+        branchName: params.branch,
+        originCommit: commit,
+        content: projectContents,
+      },
+    })
+  }
+}
+
+async function writeZipballToTempFile(params: {
+  commit: string
+  owner: string
+  repo: string
+  data: ArrayBuffer
+}): Promise<{ archiveName: string; zipFilePath: string }> {
+  const tempDir = os.tmpdir()
+
+  const shortSha = params.commit.slice(0, 7)
+  const archiveName = `${params.owner}-${params.repo}-${shortSha}`
+
+  const zipFileName = `${archiveName}-${uuid.v4()}.zip`
+  const zipFilePath = urlJoin(tempDir, zipFileName)
+
+  return new Promise((resolve, reject) => {
+    fs.writeFile(zipFilePath, Buffer.from(params.data as ArrayBuffer), (err) => {
+      if (err != null) {
+        reject(err)
+      } else {
+        resolve({
+          archiveName: archiveName,
+          zipFilePath: zipFilePath,
+        })
+      }
+    })
+  })
+}
+
+async function uploadAssets(params: { assets: AssetToUpload[]; projectId: string }): Promise<void> {
+  // Locally, store the assets on disk.
+  // On production, upload them to S3.
+  const uploads = params.assets.map(async (asset) => {
+    switch (ServerEnvironment.environment) {
+      case 'local':
+      case 'test':
+        await saveFileToDisk(params.projectId, asset)
+        break
+      case 'prod':
+      case 'stage':
+        await saveFileToS3(params.projectId, asset)
+        break
+      default:
+        assertNever(ServerEnvironment.environment)
+    }
+  })
+  await Promise.all(uploads)
+}
+
+export function unzipGithubArchive(params: {
+  archiveName: string
+  zipFilePath: string
+  existingAssets: ExistingAsset[]
+}): Promise<UnzipResult> {
+  return (
+    new Promise<UnzipResult>((resolve) => {
+      let assetsToUpload: AssetToUpload[] = []
+      let root: ProjectContentDirectory = projectContentDirectory('')
+      fs.createReadStream(params.zipFilePath)
+        // NOTE: this is parsed in-memory (instead of unzipper.Extract), but if memory becomes a problem, we can easily convert this to be disk-backed
+        .pipe(unzipper.Parse())
+        // for each entry, process it and update the project contents, as well as adding assets to upload later
+        .on('entry', processEntry(params.archiveName, root, params.existingAssets, assetsToUpload))
+        .on('finish', () => {
+          resolve({ projectContents: root.children, assetsToUpload: assetsToUpload })
+        })
+    })
+      // 5. dispose of the zip file
+      .finally(() => {
+        fs.rmSync(params.zipFilePath)
+      })
+  )
+}
+
+export function processEntry(
+  archiveName: string,
+  root: ProjectContentDirectory,
```

Comment:
For the sake of performance (and memory!) I'm fine with us using mutation here, but please can you call out the mutated params in each of these functions (we have a horrible naming convention of either adding `_MUTATE` to function names that mutate the params, or even naming the params something to signify that they will be mutated)

Reply:
done in https://github.com/concrete-utopia/utopia/pull/5639/commits/f838374cde6518189ce8d90179cf91927004a0c1